app/services/uuid_reporter.rb

```diff
@@ -0,0 +1,134 @@
+  #   3. Export a CSV file with email addresses and associated uuids
+  #
+  # @return [Integer] the number of uuids collected
+  def run
+    @agency = find_agency
+    @user_ids = collect_user_ids
+    @identity_hash = collect_identities
+    save_csv
+
+    identity_hash.length
+  end
+
+  private
+
+  attr_reader :emails, :issuers, :output, :agency, :user_ids, :identity_hash
+
+  def parse_file(file)
+    File.readlines(file, chomp: true)
+  rescue Errno::ENOENT
+    raise ArgumentError.new("#{file} does not exist")
+  end
+
+  def validate_emails
+    return if emails.all? { |e| ValidateEmail.valid?(e) }
+
+    raise ArgumentError.new('All of the email addresses must be valid emails')
+  end
+
+  def validate_issuers
+    all_issuers_belong_to_an_sp?
+    all_issuers_belong_to_same_agency?
+  end
+
+  def all_issuers_belong_to_an_sp?
+    return if ServiceProvider.where(issuer: issuers).count == issuers.length 
+
+    raise ArgumentError.new('All of the issuers must correspond to a service provider')
+  end
+
+  def all_issuers_belong_to_same_agency?
+    agency_count = Agency.
+      joins(:service_providers).
+      where(service_providers: { issuer: issuers }).
+      distinct.
+      count
+
+    return if agency_count == 1
+
+    raise ArgumentError.new('All of the issuers must belong to the same agency')
+  end
+
+  def validate_output
+    return unless File.exist?(@output)
+
+    raise ArgumentError.new('Output file already exists')
+  end
+
+  def find_agency
+    Agency.
+      joins(:service_providers).
+      find_by(service_providers: { issuer: issuers.first })
+  end
+
+  def collect_user_ids
+    emails.each_with_object({}) do |email, hash|
+      user_id = EmailAddress.confirmed.find_with_email(email)&.user_id
+      hash[email] = user_id
+    end
+  end
+
+  def collect_identities
+    # This makes use of the composite indexes on both the identities table
+    # (user_id, service_provider) and the agency_identities table (user_id,
+    # agency_id) so it should not require a full scan of either table.
+    query = AgencyIdentity.
+      select(:uuid, :user_id).
+      joins(agency: { service_providers: :identities }).
+      where(
+        identities: { user_id: user_ids.values, service_provider: issuers },
+        user_id: user_ids.values, agency_id: agency.id
+      )
+
+    # SELECT agency_identities.user_id, agency_identities.uuid
+    # FROM agency_identities
+    # INNER JOIN agencies
+    #   ON agency_identities.agency_id = agencies.id
+    # INNER JOIN service_providers
+    #   ON service_providers.agency_id = agencies.id
+    # INNER_JOIN identities
+    #   ON identities.service_provider = service_providers.issuer
+    # WHERE identities.user_id IN ?
+    #   AND identities.service_provider IN ?
+    #   AND agency_identities.user_id = ?
+    #   AND agency_identities.agency_id = ?
+
+    uuid_hash = query.each_with_object({}) do |record, hash|
+      hash[record.user_id] = record.uuid
+    end
```

Comment:
My favorite way to construct a hash is:

```suggestion
    uuid_hash = query.map { |record| [record.user_id, record.uuid ] }.to_h
```

Reply:
nice, looks cleaner for sure 😄. Updating!